pkg/tc/actuator_tc.go

```diff
@@ -0,0 +1,108 @@
+package tc
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+
+	"github.com/Mellanox/multi-networkpolicy-tc/pkg/tc/types"
+)
+
+// NewActuatorTCImpl creates a new ActuatorTCImpl
+func NewActuatorTCImpl(tcIfc TC, log klog.Logger) *ActuatorTCImpl {
+	return &ActuatorTCImpl{tcApi: tcIfc, log: log}
+}
+
+// ActuatorTCImpl is an implementation of Actuator interface using provided TC interface to apply TC objects
+type ActuatorTCImpl struct {
+	tcApi TC
+	log   klog.Logger
+}
+
+// Actuate is an implementation of Actuator interface. it applies TCObjects on the representor
+// Note: it assumes all filters are in Chain 0
+func (a *ActuatorTCImpl) Actuate(objects *TCObjects) error {
+	// list qdiscs
+	currentQDiscs, err := a.tcApi.QDiscList()
+	if err != nil {
+		return errors.Wrap(err, "failed to list qdiscs")
+	}
+
+	var ingressQDiscExist bool
+	for _, q := range currentQDiscs {
+		if q.Type() == types.QDiscIngressType {
+			ingressQDiscExist = true
+		}
+	}
+
+	if objects.QDisc == nil {
+		// delete ingress qdisc if exist
+		if ingressQDiscExist {
+			return a.tcApi.QDiscDel(types.NewIngressQdisc())
+		}
+		return nil
+	}
+
+	if len(objects.Filters) == 0 {
+		// delete filters in chain 0 if exist
+		chains, err := a.tcApi.ChainList(types.NewIngressQdisc())
+		if err != nil {
+			return err
+		}
+
+		for _, c := range chains {
+			if *c.Attrs().Chain == 0 {
+				return a.tcApi.ChainDel(objects.QDisc, types.NewChainBuilder().WithChain(0).Build())
+			}
+		}
+		return nil
+	}
+
+	// add ingress qdisc if needed
+	if !ingressQDiscExist {
+		if err = a.tcApi.QDiscAdd(objects.QDisc); err != nil {
+			return err
+		}
+	}
+
+	// get existing filters
+	existing, err := a.tcApi.FilterList(objects.QDisc)
+	if err != nil {
+		return err
+	}
+
+	// create filter sets
+	existingFilterSet := NewFilterSetImpl()
+	newFilterSet := NewFilterSetImpl()
+
+	for _, f := range existing {
+		existingFilterSet.Add(f)
+	}
+	for _, f := range objects.Filters {
+		newFilterSet.Add(f)
+	}
+
+	if existingFilterSet.In(newFilterSet) && existingFilterSet.Len() == newFilterSet.Len() {
```

Comment:
nit: add `Equal` method to FilterSet?

Reply:
ack, i also swapped the conditions to make returning false faster

Reply:
addressed in a separate PR #16